### DISCUSSION TITLE
Problems with Physics when use animation

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1545] [Q&A]
Hi,

I have a small problem when i use Physics with animation.

I made a small project to show:
https://github.com/paulocoutinhox/axmol-bomb-test

When i use this code:

```
// create animation
auto animation = Animation::createWithSpriteFrames(frames, 0.1f);
auto animate = Animate::create(animation);

auto bomb = Sprite::create();
bomb->runAction(RepeatForever::create(animate));

bomb->setName("shoot");
bomb->setContentSize(Vec2(78, 64));

// setup the physical body
auto physicsBody = PhysicsBody::createCircle(bomb->getContentSize().width / 2, PhysicsMaterial(1.0f, 0.1f, 0.8f));
physicsBody->setMass(1.0f);
bomb->setPhysicsBody(physicsBody);
```

The bomb goes to:

<img width="1238" alt="ss3" src="https://github.com/axmolengine/axmol/assets/395096/9b8497f9-2e73-4a16-93f3-fe3cde43ab05">

But when i use this code:

```
auto bomb = Sprite::create( "bomb.png");

bomb->setName("shoot");
bomb->setContentSize(Vec2(78, 64));

// setup the physical body
auto physicsBody = PhysicsBody::createCircle(bomb->getContentSize().width / 2, PhysicsMaterial(1.0f, 0.1f, 0.8f));
physicsBody->setMass(1.0f);
bomb->setPhysicsBody(physicsBody);
```

The bomb goes to down, correctly.

Thanks.

**Comment by aismann** [comment]
@paulocoutinhox 
Please have a closer look to cpp-test->42.Node: Physics->5.Actions

**Comment by paulocoutinhox** [comment]
I see, but don't have nothing related to my problem.

That test only set gravity to zero and make position actions, that is not my case.

Did you see my problem above ^?

The physics works, until my sprite is not animated, but normal sprite works.

**Comment by aismann** [comment]
Once again:
Please have a closer look to cpp-test->42.Node: Physics->5.Actions

Easiest way:
Add your code there and check behaviour again.
Maybe change the gravity too.

Whats happen?

**Comment by paulocoutinhox** [comment]
I dont want change the gravity man, i only want that the sprite be normal and not crazy, going to bottom left.

And my code is above and the sample too.

**Reply by aismann** [reply]
@paulocoutinhox 
What I mean is a cpp-test->42.Node: Physics->'Actions with SpriteFrames' is missed.
Want you add your example?

**Comment by paulocoutinhox** [comment]
To you understand better:


https://github.com/axmolengine/axmol/assets/395096/2e80e8cf-3779-47be-a226-bdcb117009f5

Player 1 - Left - Is using simple sprite
Player 2 - Right - Is using animated sprite

**Reply by aismann** [reply]
Nice.
For me its looks only the launch angle be wrong.
Maybe a deepher look whats is different with animated/normal sprite setting the angle is needed.
@paulocoutinhox 
Can you do it?

**Reply by paulocoutinhox** [reply]
the second player shoot is very crazy and go to random place.

**Reply by paulocoutinhox** [reply]
There is no difference between the code, the only difference is about the sprite creation, one have animation and other no. If i remove animation, as i sad below, it works.

**Comment by aismann** [comment]
> the second player shoot is very crazy and go to random place.

Have you looked on the sources (physics engine) about the different behaviour?

**Comment by paulocoutinhox** [comment]
Basically the problem is here:
```
sprite->runAction(RepeatForever::create(animate));
```

If i comment this line, it works.

**Comment by rh101** [comment]
Look at `Sprite::setSpriteFrame(SpriteFrame* spriteFrame)`, which is the method that is called when an `Animate` action is running on a sprite, and it seems to be the one that changes values in the sprite which may affect the physics.  

Comment out the contents of that method, and see if the physics is correct.  If it is, then uncomment a few lines in that method and re-run the app, and repeat the process until you figure out which bits of code are affecting the physics.

**Reply by aismann** [reply]
@rh101 
Maybe a Wiki entry "How can I help to find an axmol bug/issue be myself:" can improve such kind of reports too?

**Reply by paulocoutinhox** [reply]
Hi @rh101, i already debug it, but has nothing.

Can be something related to rotation. Something internally can be changing rotation.

Because i not found that im trying get some help.

**Reply by paulocoutinhox** [reply]
By the logic of the thing, whether the sprite is animated or not, the physics (or whatever) should work the same, in the same way.

**Reply by paulocoutinhox** [reply]
I use sprites in other engines like unity/godot and it work with physics as the same way, with or without animation. To me it is a internal bug of cocos2dx/axmol that make action change sprite behavior. A sprite cannot go to side left when you animate it.

**Comment by aismann** [comment]
> Because i not found that im trying get some help.

I think I can speak for all 'Top 10' contributors:
We try to fix as fast as possible.
- But we make it for free and mostly for having fun.
- But each detected bug/issue can also be on a roadmap and not fixed as fast as possible.

Off topic:
Maybe a small donation is a possisble way to push some whishes on the top. 
But this is not realy easy to handle and not a good way.
I have seen such requests on lot of 'open source' frameworks, also on cocos2d-x too.

**Reply by paulocoutinhox** [reply]
But I'm not demanding anything, I just asked for help. Help me with your time. The most obvious things I have already validated.

**Reply by rh101** [reply]
> Help me with your time.

The most valuable thing in the world. ;)

As @aismann was alluding to, this isn't anyone's job, so there can be no expectation placed on people to answer questions in a timely manner, or even answer them at all. 

The only thing I've noticed is that a person can get a lot more help if they have shown that a decent amount of effort has been put into tracking down the source of the issue (simply debugging may not be enough).

What you know so far is that the issue occurs when the sprite is animated.  Debugging isn't going to help in this case, because you need to see what values are being generated (rotation, position etc. etc.).  Using your game as the example here, set up 2 cannons, both facing the exact same way, and in exactly the same position.  One cannon has an animating sprite, the other non-animating sprite.  Trigger the cannon fire and ensure you are logging all relevant data (rotation, position etc.) to the console for each of the cannons.  Once you've got the data, compare to see what the differences are.  That may help you figure out what the problem is.

**Reply by paulocoutinhox** [reply]
I made that demo on github (axmol-bomb-test) precisely to show the problem, with a very simple code so anyone can help and see the problem occurring. It seems to me something very common and simple that anyone could have this problem, if they use an animated sprite with physics enabled.

**Reply by aismann** [reply]
@paulocoutinhox 
At the moment its (only) a known bug. Roadmap has to be create for this issue maybe.
If you can't wait?
Please start what @rh101 has written: As more info and stuffs be done as more help will you get or at least learn  helping to fix it. And last but not least: please hold in mind: Nobody have to helping you but may later or may now.

**Comment by paulocoutinhox** [comment]
The other strange thing is that the problem only happens on desktop, because on appletv and ios it works without problem. See: 

https://github.com/axmolengine/axmol/assets/395096/59176bfb-bd98-4b7b-875b-ef8b036d25d1

**Reply by paulocoutinhox** [reply]
Can anyone help me with this?

**Comment by paulocoutinhox** [comment]
Hi,

Can anyone help me with this?

Thanks.

**Comment by iAndyHD3** [comment]
You just need more testing to try to find the problem

1. Are you sure it breaks with any animation?
2. Try other animation to see if anything changes
3. Try stoping and resuming the animation multiple times and see if anything changes
4.  Try calling the [Animation::setLoops](https://axmolengine.github.io/manual/latest/d5/d3f/classax_1_1_animation.html#a2d39129f5b5a48287d0e7219fbffa801) method instead of using `RepeatForever`
5. Finally, if none of the above fixes it, try making a manual animation calling `Sprite::setSpriteFrame` by yourself.

**Comment by aismann** [comment]
@paulocoutinhox 
You can close this here?
